test: Http2Stream redundant call to shutdown and call passing only callback

trivikr opened the pull request:
This PR includes test cases for the handling of:
- redundant call of `stream.session.shutdown()`
- calling `stream.session.shutdown()` with only callback (no options)

Refs: #14985

##### Checklist
- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
test, http2

apapirovski commented on test/parallel/test-http2-server-shutdown-redundant.js:
```diff
@@ -0,0 +1,28 @@
+  stream.session.shutdown(common.mustCall(() => {
+    assert.strictEqual(
+      stream.session.shutdown(),
+      undefined
```
I think it's better if this also checks that it `doesNotThrow`.

trivikr replied:
`assert.strictEqual` will ensure that no error is thrown
Tested on [repl](https://repl.it/LaoE/1)

apapirovski commented on test/parallel/test-http2-server-shutdown-redundant.js:
```diff
@@ -0,0 +1,28 @@
+server.on('stream', common.mustCall((stream) => {
+  stream.session.shutdown(common.mustCall(() => {
+    assert.strictEqual(
+      stream.session.shutdown(),
```
I would suggest adding `common.mustNotCall()` so that it's clear that the callback only executes once.

trivikr replied:
Added!

apapirovski commented on test/parallel/test-http2-server-shutdown-redundant.js:
```diff
@@ -0,0 +1,28 @@
+
+// Test blank return when a stream.session.shutdown is called twice
+server.on('stream', common.mustCall((stream) => {
+  stream.session.shutdown(common.mustCall(() => {
```
I would also run another `stream.session.shutdown(common.mustNotCall())` outside of the callback. If there are two calls in a row, it shouldn't throw or execute a callback.

trivikr replied:
Added!

trivikr commented:
I agree to comment by @apapirovski https://github.com/nodejs/node/pull/15612#pullrequestreview-64962525 that the test test-http2-server-shutdown-redundant.js is a subset of test in test-http2-server-shutdown-only-callback.js
I've kept them separate as they serve specific purpose.

The test to cover `this[kState].destroyed || this[kState].destroying` was added in my first PR with NodeJS at https://github.com/nodejs/node/pull/15597

apapirovski commented:
Yeah, my bad re: the 2nd point. 

I'm still not sure I understand what `test-http2-server-shutdown-only-callback.js` does that the other test doesn't. It seems like a naming and comment problem rather than a distinct test.

trivikr commented:
I've removed the test `test-http2-server-shutdown-only-callback.js` and added a comment in `test-http2-server-shutdown-redundant.js` that it also tests only callback use case.

trivikr commented:
When the commits are squashed before merge, the commit message should be:

```
test: Http2Stream redundant shutdown and single callback

Refs: #14985
```

apapirovski commented:
Thanks @trivikr! 👍 I realize these were sort of nits but just to explain my reasoning: 

There are currently over 1200+ tests and it's growing all the time, so as much as possible it's good to avoid highly specialized test cases if the same can be accomplished by a single slightly longer test case that goes through a few stages (as long as it doesn't compromise the test result). That way we can avoid having duplicate code and unnecessary amount of buildup and teardown of tests. (If you look at the node CI, it runs all the tests on some truly ancient/slow systems.)

See some of these other http2 tests:
https://github.com/nodejs/node/blob/master/test/parallel/test-http2-compat-serverresponse-end.js
https://github.com/nodejs/node/blob/master/test/parallel/test-http2-compat-serverresponse-destroy.js

trivikr commented:
I completely agree with @apapirovski 
The command `make test` takes around 20 minutes to run on MacBook Pro (Retina, 13-inch, Early 2015)

It's difficult to run single test also. I went through Makefile and came up with a python command to runs single test https://github.com/nodejs/node/blob/4f4d55d10e24d2ae31103527cc380f359f34173f/Makefile#L204

Questions:
- Is easing development by providing an ability to debug/run single tests being tracked somewhere? If the support is already available, can it be updated in [writing tests guide](https://github.com/nodejs/node/blob/master/doc/guides/writing-tests.md)
- Does NodeJS plan to move a easier/faster test framework like [Jest](https://facebook.github.io/jest/)?

The above issues weren't blocking me for this task, so I didn't create issues.
We can create issues/tasks and follow up to get more developers onboard.

bnoordhuis commented:
> Is easing development by providing an ability to debug/run single tests being tracked somewhere?

It's documented in CONTRIBUTING.md, if that is what you mean.  You are welcome to add it to the 'writing tests' guide though.

> Does NodeJS plan to move a easier/faster test framework like Jest?

No.

trivikr commented:
Squashed all commits into a single one, and force pushed it my private branch.

trivikr commented:
What's the procedure to merge these changes to master? Do we have wait for certain time for approvals before merging?

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/10308/

trivikr commented:
@BridgeAR rebased with https://github.com/nodejs/node/tree/master

jasnell commented:
@trivikr ... yes, PRs generally must sit for at least 48 hours during the week and 72 hours over the weekend before they can be landed. Exception is given only to trivial doc updates. Also, things must go through a CI test run.

jasnell commented:
Oh, and I see you added a merge commit. Generally, we always use rebase and avoid using merge commits. Before I get this landed, can I ask you to drop the merge commit and rebase from master?

trivikr commented:
Dropped the merge commit and performed git rebase as suggested by @jasnell and [documentation](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#step-9-discuss-and-update)

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/10354/

BridgeAR commented:
Landed in 27d8202db54ce3eda6ff47703169e577476c9dd2

trivikr commented:
Thanks @BridgeAR 
This is my first PR to NodeJS - a small step to contribute more to this awesome project :-)

apapirovski commented:
Congrats @trivikr! Thanks for contributing 🎉
